### ISSUE TITLE
Edges become invalid after related collections get reimported

### ISSUE THREAD

**Issue by razvanphp** [arangodb/arangodb#1581]
I hope the title is self explanatory, but this is really a pain in the a**, last night I wanted to keep a very big edge collection, that joins customers with posts, but reimport from a dump the latest customers and posts from `arangodump`.

Everything seem fine, but the edge became invalid, so all `_to` and `_from` were pointing to `_unknown/12345`. Unloading and loading the collection does not help, also restarting arango.

What worked in the end and took few hours, was dumping the edge again, sed the data file to replace the invalid pointers and reimport:
```
sed -i 's#"_to":"_unknown#"_to":"posts#g' customerCanSeePosts_185898d7669bb4264e3d00c0fa264afe.data.json
sed -i 's#"_from":"_unknown#"_from":"customers#g' customerCanSeePosts_185898d7669bb4264e3d00c0fa264afe.data.json
```

This is a huge pain when dealing with big data, so maybe you can improve this somehow.

Thank you!

**Comment by neunhoef** [comment]
Sorry for the inconvenience. Let me first explain why the setup is as it is: The edges store the vertex collection*id* rather than the collection name in the _from and _to fields. This is intentional to allow renaming a vertex collection. If one renames a vertex collection, all _from and _to entries in the edge collections are changed automatically!

That is why the edges had the "_unknown" entries when you had dropped the vertex collection. When you then used `arangorestore` to recreate the vertex collection, the new one had the same name but a different (internal) ID, and so the edges still had "_unknown". Your fix with dump/restore with a sed-step was one sensible approach to fix the mess.

What you could have done is to use `truncate()` rather than `drop()`. This would have deleted all vertices but would have retained the collection with its ID. The edges at this stage would still have been happy! Then you could have used

    arangorestore --collection V --create-collection false --input-directory dump

(assuming your vertex collection is `V` and `dump` contains the dump of your database). And everything would have been fine.

I admit that this is totally unintuitive and has the additional disadvantage that `truncate()` is slower than `drop()`, since it has transactional behaviour (can be rolled back when issued within a transaction).
 
We have concluded that we would like to improve the following things in one of the next versions:

  - make vertex collections aware of the edge collections in which they are referenced and prevent the drop() call in that case. This could for example work by forcing users to declare with each edge collection the possible vertex collections occurring in the _from and _to fields. Some more design work and discussion are needed to finalize this idea.
  - create a `truncateFast` call which is as fast as `drop()` but retains the collection with its ID
  - find a good place to document this issue better

**Comment by razvanphp** [comment]
I totally understand your explanation, it makes sense, and I understood this is the case when I stumbled upon this last night.

What I would suggest, and I think it's a simpler solution, is to allow changing the internal id of the collection, from the web interface, if this does not break anything. Or maybe, arangodump should include the internal id in the dump, so when he drops the collection and creates a new one he still keeps the same id, I think this is expected anyway, I don't see any reason to change the id if you drop anyway the old collection.

**Comment by neunhoef** [comment]
We have to be very careful here indeed. If a collection is dropped it is entirely possible that some JavaScript stuff or at least some markers in the write ahead log still mention the old collection. Asynchronous replication might still work with the old collection! There might even be a data directory for it on disk which we need to hold until all references to the collection have vanished. All of these references, directory names etc. contain the collection ID rather than the name.
Therefore, recreating a new collection with the same ID in a running server is a very dangerous thing to do, because the old references might now believe that they mean the new collection! It could even happen that the creation does not work, since the old data directory is still present. Nevertheless we do want to allow the user to recreate a collection with the same name, of course.
`arangodump` does indeed dump the name and the ID, but in its standard mode `arangorestore` restores the "same" collection with a different ID. This is usually OK, except in the case of edges mentioning the ID.
`arangorestore` even allows the option `--recycle-ids true` which does indeed use the old ID. But as mentioned above this is dangerous and should only be used in a newly created `arangod`.

**Comment by razvanphp** [comment]
Ok, then what about the other way around, editing the edge collection to fix the missing id references?

For example on the graph definition, there is a possibility to choose the edge start and end points, and after it is set, the web interface shows only the edge colletion name. Maybe when you detect that it is `_unknown` you can allow similarly to redefine the new vertex id, along with maybe unloading and loading the edge again.

**Comment by neunhoef** [comment]
We could offer an API to "repair" an edge collection. One would have to specify the mapping from old collection IDs to new ones. This could be done in a reasonably fast fashion. It would however involve writing a new revision for each edge in the collection. So, if one knows in advance that one wants to completely rewrite a vertex collection, going the truncate() and restore option is still better.

**Comment by razvanphp** [comment]
Maybe it makes no sense what I'm saying, but why every document in the edge collection references the vertex collection id? Wouldn't make more sense that this is an attribute of the edge collection itself? This way you would not have to update each edge for this and write a new revision.

**Comment by jsteemann** [comment]
This would be ideal when all edges in a collection reference the exact same vertex collections in their `_from` and `_to` attributes.
This is indeed one use case, but in general an edge collection can refer to any number of collections. 

Saving the referred collections per edge will work in this case, and saving it per collection would be problematic here, as different edges may be referring to completely different collections. 

I've seen this going from just 1 referred-to collection up to around 20 different collections referred to from the same edge collection. From my point of view, this would allow for storing some small translation dictionary for the edge collection, which could provide an id for a collection name. These ids would be valid inside the edge collection only, and would not be the same as collection ids. The advantage of this would be that these ids would still be valid after a referred-to-collection is dropped, because the translation dictionary would still be there. When a collection with the same name gets re-created everything would be fine still. When a collection gets renamed, one would need to change the collection name in the dictionary, but would not need to touch all edges.

**Comment by wirtsi** [comment]
So is there any progress on this? 

Because what this boils down to is that when my edge collection really references various collections, then I have no way of doing a restore.

This would leave me in a pretty hairy situation should the need arise to do a restore!

**Comment by dothebart** [comment]
@wirtsi you will only get into this situation, if you actually **drop referenced vertex collections** right *before* doing a dump. For regular backups should dump your whole database, then this problem won't bite you.

**Comment by wirtsi** [comment]
Thanks for the clarification ... so it is not as bad as I had feared. Still it is a bit odd, because I did not actually drop any referenced tables but did a partial dump. This seems to trigger the same behaviour

So something like arangodump --collection posts --collection customers --collection edge

**Comment by dothebart** [comment]
Its not related to not dumping all collections. Its related to that you dropped and recreated the vertex collection meanwhile. So if you dump and restore your vertex collection only, you will get into this.

**Comment by razvanphp** [comment]
Ok, so the situation arised again, and problem is `--recycle-ids=true` for `arangorestore` does not have the desired effect, the edges still get invalid.

What I did was to use arangodump with 3 collection names (without the edge) and then use arangorestore with the parameter. Is there any way I can prevent the drop and recreate of the collections using this way? How do I make him use truncate instead?

**Comment by jsteemann** [comment]
I still arangorestore will not import edges that link to non-existing collections.
Non-existing here means that the linked collection does not exist on the server and will not be created by the restore.
I think one can tell arangorestore to go on in case an error occurs, using the `--force` option. That will silently discard any edges/documents that cannot be imported, but should import those that do not cause problems.

By the way, in 3.0 the handling of edges is fixed as follow:
- `_from` and `_to` will store collection names, not collection ids. This way dropping a collection that is linked by any `_from` or `_to` is not problematic if it is re-created later. The edges stay intact and keep their meanings. There is no need to re-create collections with the same ids, as everything will work by name then.
- arangodump will export the edges as they are. As `_from` and `_to` are string values in 3.0, their values will be exported, regardless of whether the linked collections exist or not.

Here's an example of how edges stay intact, even if a linked collection is dropped:
```
/* set up two vertex collections and an edge collection that links them */
arangod> db._create("vertex1");
arangod> db._create("vertex2");
arangod> db._createEdgeCollection("edges");
arangod> db.edges.insert({ _from: "vertex1/1", _to: "vertex2/1" });
{ 
  "_id" : "edges/3793813", 
  "_key" : "3793813", 
  "_rev" : "3793813" 
}
arangod> db.edges.any()
{ 
  "_key" : "3793813", 
  "_id" : "edges/3793813", 
  "_from" : "vertex1/1", 
  "_to" : "vertex2/1", 
  "_rev" : "3793813" 
}

/* now drop one of the linked collections */
arangod> db._drop("vertex1");

/* edge still intact */
arangod> db.edges.any()
{ 
  "_key" : "3793813", 
  "_id" : "edges/3793813", 
  "_from" : "vertex1/1", 
  "_to" : "vertex2/1", 
  "_rev" : "3793813" 
}
```

**Comment by razvanphp** [comment]
Nice, I thought this was not possible, seems you found a solution.

In my case it worked also in version 2, as I imported only the vertexes, not edges, and my goal was that existing edges don't get corrupted.

Looking forward to version 3 with this and persistent indices.

**Comment by jsteemann** [comment]
Yes, we found a solution. We tried out a few variants and the only robust solution was switching from storing edge collection ids to storing edge collection names.
Obviously we weren't able to do this change in the stable 2.8 branch, so but the change has been in devel for quite a while. We're also keen to get 3.0 out. A think the RC version is going to be released soon.

**Comment by dothebart** [comment]
We just released the ArangoDB 3.0 RC2 with fix for this issue: https://www.arangodb.com/download-preview/
